README.md
# [Baka MPlayer](http://bakamplayer.u8sand.net)

[![Build Status](https://travis-ci.org/u8sand/Baka-MPlayer.svg?branch=master)](https://travis-ci.org/u8sand/Baka-MPlayer)

--------------


## Overview

Baka MPlayer is a free and open source, cross-platform, **libmpv** based multimedia player.
Its simple design reflects the idea for an uncluttered, simple, and enjoyable environment for watching tv shows.


## Requirements

* gcc
* pkg-config
* libmpv-dev
* qtbase5-dev (>= 5.2.0)
  * qt5-qmake
  * qttools5-dev-tools
  * qtdeclarative5-dev
  * libqt5svg5-dev
  * libqt5x11extras5-dev
  * libqt5network5
* youtube-dl (optional, for streaming youtube videos)

Note: Packages may be named slightly different for each distro

### Get the font

Baka MPlayer was designed around the font called Noto Sans. Noto Sans was used because of its open source nature and its broad support for Unicode characters. Having the correct font installed insures that what you see is what was intended.

[Get it here.](https://www.google.com/get/noto/#sans-lgc)


## Compilation

### Windows

These instructions are for cross-compiling for Windows on a Linux system (Note: the architecture can be either `x86_64` or `i686` depending on which platform you're compiling for).
```
git clone -b release https://github.com/u8sand/Baka-MPlayer.git
cd "Baka-MPlayer"
mkdir build
cp -r windows/cross-compilation/* build/
cd build
arch=x86_64
./baka-build.sh $arch
```
This is a very long process because you'll need to build the mingw32 toolchain `mxe` and all dependent libraries, `libmpv.a`, and finally `baka-mplayer.exe`. If everything succeeded without error, you'll get `Baka-MPlayer.$arch.zip` which should contain everything you need.

To rebuild simply delete the directory (in build) or the .zip file of what you need to rebuild and re-run `./baka-build.sh $arch`.

To add custom patches, put them in `src/patches/` prefixed with the name of what you're patching.

### Linux

If your distribution does not provide a package, you can compile it from source.
However we've made scripts for some distributions. See `etc/sbin/linux/`.
```
git clone -b release https://github.com/u8sand/Baka-MPlayer.git
cd "Baka-MPlayer"
mkdir build
cp -r linux/* build/
cd build
distro=debian_based
./$distro.sh
```
If this doesn't work or the distro you use is not listed here, you'll need to build mpv and then Baka MPlayer (the dependencies above are for Baka MPlayer). For help building mpv see `https://github.com/mpv-player/mpv-build`. Compiling Baka MPlayer from source can be done like so:
```
git clone -b release https://github.com/u8sand/Baka-MPlayer.git
cd "Baka-MPlayer"
./configure
make -j `grep -c ^processor /proc/cpuinfo`
sudo make install
```
The configuration file will be created on first run and will be written to `~/.config/bakamplayer.ini`.

### macOS

As of right now you have to compile from source.
You require some more dependencies which you can get from homebrew:
```
brew install youtube-dl
brew install mpv --with-libmpv
git clone -b release https://github.com/u8sand/Baka-MPlayer.git
cd Baka-MPlayer
export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig
brew install qt5
export QMAKE=/usr/local/Cellar/qt5/5*/bin/qmake
./configure CONFIG+=install_translations
make
open build
```
Then drag the app to your Applications folder.
In the future there will be a brew formula to simplify this.

Similarly to Linux, the configuration file will be created on first run and will be written to `~/.config/bakamplayer.ini`.

### Other languages

By default, Baka MPlayer will compile in English if no language is specified during compilation. To compile a multi-lingual version of baka-mplayer, configure it like so:

    ./configure CONFIG+=install_translations

For more configuration options see the `configure` source file or read the manual.

You can check out which languages we currently support by checking out `Baka-MPlayer/src/translations/`.


## Bug reports

Please use the [issues tracker](https://github.com/u8sand/Baka-MPlayer/issues) provided by GitHub to send us bug reports or feature requests.


## Contact

**IRC Channel**: `#baka-mplayer` on `irc.freenode.net`

You can ask us questions about using Baka MPlayer, give feedback, or discuss its development.
However, if possible, please avoid posting bugs there and use the [issue tracker](https://github.com/u8sand/Baka-MPlayer/issues) instead.

## Windows 下编译出中文版本
需要在qmake处增加额外的参数："CONFIG+=embed_translations"


debian/changelog
baka-mplayer (3.0.0-1) unstable; urgency=low
  
  * New: 始终打开一个播放实例，并立即激活之前的窗口
  * Fixed: 窗口标题，播放某些文件名显示乱码
  * Changed: 拖动进度条的时候，显示的时间有缓慢滑入滑出的效果，产生延迟感，故直接剔除
  * Changed: 更换播放器的图标
  * Changed: 去掉打开另外一个播放器的菜单
  * Changed: 去掉在线帮助的菜单
  * Changed: 去掉关于页面下的乱码
  * Changed: 去掉检查更新的菜单和功能，调整版本号为3.0.0
  * Changed: 不显示播放列表
  * Changed: 第二个播放剩余时间，改为总时长
  
 -- wankaiming 2020.3.14

baka-mplayer (2.0.4-1) unstable; urgency=low

  * New: Highly requested Hide All Controls mode (under Settings menu)
  * New: Ability to add external audio files
  * New: Option to not save recently opened files
  * New: Option to not restore playback position
  * New: Option to change mpv msg-level
  * New: "clear" command
  * New: Explain how to leave full screen mode
  * New: Message is displayed for Motion Interpolation & Deinterlace options
  * New: Media Info displays if hwdec is in use or not
  * Fixed: Reordering playlist items would make them disappear (Windows only)
  * Fixed: Show message if file couldn't be opened
  * Fixed: An issue with resuming playback position
  * Fixed: Hide Album Art fully hides that portion
  * Fixed: Command line output is no longer double spaced
  * Changed: Double click a Key Binding entry to change it
  * Changed: Renamed an item to Open Copied Link
  * Changed: Cmd line shortcut key is now Ctrl+`
  * Changed: Cmd line no longer requires "debug debug" to show output
  * Changed: Gesture info is visually represented
  * Changed: Button text case to Sane Case

 -- Joshua Park (godly-devotion) <azian.otaku@live.com>  Tue, 1 Dec 2015 08:00:00 -0500

baka-mplayer (2.0.3-1) unstable; urgency=low

  * Hint: If you speed up the video, Baka will display how much real time you have remaining.
  * Fixed: Status info is now displayed with audio files without album art
  * Fixed: Disable Show in Folder for online files
  * Fixed: Key Bindings table is, and can be, sorted alphabetically
  * Changed: Logo :D
  * Changed: Settings file uses JSON
  * Changed: Currently playing file is selected in playlist (old behavior)
  * Changed: External subtitles will be labeled as so
  * Changed: Volume level shown on mouse scroll
  * Changed: Speed is increased/decreased by interval of 10%
  * Changed: Sub font size is increased/decreased by 0.1
  * Improved: Overall design
  * Improved: Playback Info display
  * Added: Croatian translation
  * Added: Dutch translation
  * Added: French translation
  * Added: Spanish translation
  * Added: Mute (different from setting volume to 0)
  * Added: Thumbnail buttons (Windows only)
  * Added: Metadata (ex. ID3 Tags) to playback info
  * Added: Delete external sub file along with video file (optional)
  * Added: Motion Interpolation option (previously called "smoothmotion")
  * Added: "help" command

 -- Joshua Park (godly-devotion) <azian.otaku@live.com>  Wed, 3 Jun 2015 16:00:00 -0500

baka-mplayer (2.0.2-1) unstable; urgency=low

  * Fixed: Dim Desktop on multiple platforms
  * Fixed: Playing files located in the root directory (Windows only)
  * Fixed: Show parent folder under recently opened list (Windows only)
  * Fixed: Show in Folder selects file (Windows only)
  * Fixed: Gesture sensitivity is resolution independent
  * Changed: Custom settings implementation
  * Changed: File info shows up as overlay
  * Changed: Some default key-bindings for easier workflow
  * Improved: Major code refracting for efficiency
  * Improved: Playlist
  * Improved: Fit Window
  * Improved: Settings
  * Improved: Multiple-monitor support
  * Improved: Tweaked interface colors for better visibility
  * Added: Italian translation
  * Added: Support for custom keyboard shortcuts
  * Added: Playlist will auto show if cursor is near it in full screen
  * Added: Automatic Updating (Windows only)
  * Added: Command-Line engine
  * Added: Support for easily updating youtube-dl

 -- Joshua Park (godly-devotion) <azian.otaku@live.com>  Sun, 25 Jan 2015 22:42:00 -0500

baka-mplayer (2.0.1-1) unstable; urgency=low

  * Windows Build Available!
  * Fixed: Spacing in debug window output
  * Fixed: Some issue with loading mpv settings
  * Fixed: Modified PKGBUILD to download much faster than before
  * Fixed: Desktop file for better platform integration
  * Fixed: Playlist search is now case-insensitive
  * Changed: Build system from CMake to QMake
  * Changed: Some menu item shortcut keys
  * Changed: Increased seekbar height
  * Changed: Fast seeking
  * Changed: Do not save window size into settings
  * Changed: Currently playing item is bolded
  * Added: Youtube video playback support (and others)
  * Added: Ability to open directories
  * Added: Recently opened file list
  * Added: Interact with mpv directly via command line interface
  * Added: Fit Window 150%
  * Added: Gesture Support
  * Added: Up/Down Arrow keys navigate playlist, Enter plays
  * Added Translations (Chinese, Korean, Portuguese, Russian)

 -- Joshua Park (godly-devotion) <azian.otaku@live.com>  Sun, 28 Dec 2014 20:30:00 -0500

baka-mplayer (2.0.0-1) unstable; urgency=low

  * Entire new look and feel
  * More complete manual
  * More efficient preferences functionality
  * Changes to right-click (now opens context menu in full  *screen mode)
  * Playlist drag and drop support
  * Mpv settings in settings file
  * Settings version information for better version upgrading
  * Baka MPlayer album art for music without album art
  * More useful Media Info dialog
  * Playback speed adjusting
  * Use of mpv OSD to notify you about various actions
  * Update manager version information
  * Window fitting probably works better
  * Screenshot Dialog for selecting location and screenshot template

 -- Joshua Park (godly-devotion) <azian.otaku@live.com>  Thu, 20 Nov 2014 11:30:00 -0500

baka-mplayer (1.9.9-1) unstable; urgency=low

  * New build system
  * Desktop file for appearing in linux distro's application menus

src/bakaengine.h
#ifndef BAKAENGINE_H
#define BAKAENGINE_H

#include <QObject>
#include <QString>
#include <QStringList>
#include <QHash>
#include <QSystemTrayIcon>
#include <QMenu>
#include <QTranslator>

class MainWindow;
class MpvHandler;
class Settings;
class GestureHandler;
class OverlayHandler;
class UpdateManager;
class DimDialog;

class BakaEngine : public QObject
{
    Q_OBJECT
public:
    explicit BakaEngine(QObject *parent = 0);
    ~BakaEngine();

    MainWindow     *window;
    MpvHandler     *mpv;
    Settings       *settings;
    GestureHandler *gesture;
    OverlayHandler *overlay;
    UpdateManager  *update;
    DimDialog      *dimDialog;

    QSystemTrayIcon *sysTrayIcon;
    QMenu           *trayIconMenu;

    QTranslator     *translator,
                    *qtTranslator;

    // input hash-table provides O(1) input-command lookups
    QHash<QString, QPair<QString, QString>> input; // [shortcut] = QPair<command, comment>

    // the following are the default input bindings
    // they are loaded into the input before parsing the settings file
    // when saving the settings file we don't save things that appear here
    const QHash<QString, QPair<QString, QString>> default_input = {
        {"Ctrl++",          {"mpv add sub-scale +0.1",              tr("Increase sub size")}},
        {"Ctrl+-",          {"mpv add sub-scale -0.1",              tr("Decrease sub size")}},
        {"Ctrl+W",          {"mpv cycle sub-visibility",            tr("Toggle subtitle visibility")}},
        {"Ctrl+R",          {"mpv set time-pos 0",                  tr("Restart playback")}},
        {"PgDown",          {"mpv add chapter +1",                  tr("Go to next chapter")}},
        {"PgUp",            {"mpv add chapter -1",                  tr("Go to previous chapter")}},
        {"Right",           {"mpv seek +5",                         tr("Seek forwards by %0 sec").arg("5")}},
        {"Left",            {"mpv seek -5",                         tr("Seek backwards by %0 sec").arg("5")}},
        {"Shift+Left",      {"mpv frame_back_step",                 tr("Frame step backwards")}},
        {"Shift+Right",     {"mpv frame_step",                      tr("Frame step")}},
        {"Ctrl+M",          {"mute",                                tr("Toggle mute audio")}},
        {"Ctrl+T",          {"screenshot subtitles",                tr("Take screenshot with subtitles")}},
        {"Ctrl+Shift+T",    {"screenshot",                          tr("Take screenshot without subtitles")}},
        {"Ctrl+Down",       {"volume -5",                           tr("Decrease volume")}},
        {"Ctrl+Up",         {"volume +5",                           tr("Increase volume")}},
        {"Ctrl+Shift+Up",   {"speed +0.1",                          tr("Increase playback speed by %0").arg("10")}},
        {"Ctrl+Shift+Down", {"speed -0.1",                          tr("Decrease playback speed by %0").arg("10")}},
        {"Ctrl+Shift+R",    {"speed 1.0",                           tr("Reset speed")}},
        {"Alt+Return",      {"fullscreen",                          tr("Toggle fullscreen")}},
        {"Ctrl+H",          {"hide_all_controls",                   tr("Toggle hide all controls mode")}},
        {"Ctrl+D",          {"dim",                                 tr("Dim lights")}},
        {"Ctrl+E",          {"show_in_folder",                      tr("Show the file in its folder")}},
        {"Tab",             {"media_info",                          tr("View media information")}},
        {"Ctrl+J",          {"jump",                                tr("Show jump to time dialog")}},
        {"Ctrl+N",          {"new",                                 tr("Open a new window")}},
        {"Ctrl+O",          {"open",                                tr("Show open file dialog")}},
        {"Ctrl+Q",          {"quit",                                tr("Quit")}},
        {"Ctrl+Right",      {"playlist play +1",                    tr("Play next file")}},
        {"Ctrl+Left",       {"playlist play -1",                    tr("Play previous file")}},
        {"Ctrl+S",          {"stop",                                tr("Stop playback")}},
        {"Ctrl+U",          {"open_location",                       tr("Show location dialog")}},
        {"Ctrl+V",          {"open_clipboard",                      tr("Open clipboard location")}},
        {"Ctrl+F",          {"playlist toggle",                     tr("Toggle playlist visibility")}},
        {"Ctrl+Z",          {"open_recent 0",                       tr("Open the last played file")}},
        {"Ctrl+`",          {"output",                              tr("Access command-line")}},
        {"F1",              {"online_help",                         tr("Launch online help")}},
        {"Space",           {"play_pause",                          tr("Play/Pause")}},
        {"Alt+1",           {"fitwindow",                           tr("Fit the window to the video")}},
        {"Alt+2",           {"fitwindow 50",                        tr("Fit window to %0%").arg("50")}},
        {"Alt+3",           {"fitwindow 75",                        tr("Fit window to %0%").arg("75")}},
        {"Alt+4",           {"fitwindow 100",                       tr("Fit window to %0%").arg("100")}},
        {"Alt+5",           {"fitwindow 150",                       tr("Fit window to %0%").arg("150")}},
        {"Alt+6",           {"fitwindow 200",                       tr("Fit window to %0%").arg("200")}},
        {"Esc",             {"boss",                                tr("Boss key")}},
        {"Down",            {"playlist select +1",                  tr("Select next file on playlist")}},
        {"Up",              {"playlist select -1",                  tr("Select previous file on playlist")}},
        {"Return",          {"playlist play",                       tr("Play selected file on playlist")}},
        {"Del",             {"playlist remove",                     tr("Remove selected file from playlist")}}
    };

public slots:
    void LoadSettings();
    void SaveSettings();

    void Command(QString command);

protected slots:
    // Utility functions
    void Print(QString what, QString who = "baka");
    void PrintLn(QString what, QString who = "baka");
    void InvalidCommand(QString);
    void InvalidParameter(QString);
    void RequiresParameters(QString);

    // Settings Loading
    void Load2_0_0();

signals:


private:
    // This is a baka-command hashtable initialized below
    //  by using a hash-table -> function pointer we acheive O(1) function lookups
    // Format: void BakaCommand(QStringList args)
    // See bakacommands.cpp for function definitions

    // todo: write advanced information about commands
    typedef void(BakaEngine::*BakaCommandFPtr)(QStringList&);
    const QHash<QString, QPair<BakaCommandFPtr, QStringList>> BakaCommandMap = {
        {"mpv", // command
         {&BakaEngine::BakaMpv,
          {
           // params     description
           QString(), tr("executes mpv command"),
           QString() // advanced
          }
         }
        },
        {"sh",
         {&BakaEngine::BakaSh,
          {
           QString(), tr("executes system shell command"),
           QString()
          }
         }
        },
        {"new",
         {&BakaEngine::BakaNew, // function pointer to command functionality
          {
           // params     description
           QString(), tr("creates a new instance of baka-mplayer"),
           QString()
          }
         }
        },
        {"open_location",
         {&BakaEngine::BakaOpenLocation,
          {
           QString(),
           tr("shows the open location dialog"),
           QString()
          }
         }
        },
        {"open_clipboard",
         {&BakaEngine::BakaOpenClipboard,
          {
           QString(),
           tr("opens the clipboard"),
           QString()
          }
         }
        },
        {"show_in_folder",
         {&BakaEngine::BakaShowInFolder,
          {
           QString(),
           tr("shows the current file in folder"),
           QString()
          }
         }
        },
        {"add_subtitles",
         {&BakaEngine::BakaAddSubtitles,
          {
           QString(),
           tr("add subtitles dialog"),
           QString()
          }
         }
        },
        {"add_audio",
         {&BakaEngine::BakaAddAudio,
          {
           QString(),
           tr("add audio track dialog"),
           QString()
          }
         }
        },
        {"screenshot",
         {&BakaEngine::BakaScreenshot,
          {
           tr("[subs]"),
           tr("take a screenshot (with subtitles if specified)"),
           QString()
          }
         }
        },
        {"media_info",
         {&BakaEngine::BakaMediaInfo,
          {
           QString(),
           tr("toggles media info display"),
           QString()
          }
         }
        },
        {"stop",
         {&BakaEngine::BakaStop,
          {
           QString(),
           tr("stops the current playback"),
           QString()
          }
         }
        },
        {"playlist",
         {&BakaEngine::BakaPlaylist,
          {
           "[...]",
           tr("playlist options"),
           QString()
          }
         }
        },
        {"jump",
         {&BakaEngine::BakaJump,
          {
           QString(),
           tr("opens jump dialog"),
           QString()
          }
         }
        },
        {"dim",
         {&BakaEngine::BakaDim,
          {
           QString(),
           tr("toggles dim desktop"),
           QString()
          }
         }
        },
        {"output",
         {&BakaEngine::BakaOutput,
          {
           QString(),
           tr("toggles output textbox"),
           QString()
          }
         }
        },
        {"preferences",
         {&BakaEngine::BakaPreferences,
          {
           QString(),
           tr("opens preferences dialog"),
           QString()
          }
         }
        },
        {"online_help",
         {&BakaEngine::BakaOnlineHelp,
          {
           QString(),
           tr("launches online help"),
           QString()
          }
         }
        },
        {"update",
         {&BakaEngine::BakaUpdate,
          {
           QString(),
           tr("opens the update dialog or updates youtube-dl"),
           QString()
          }
         }
        },
        {"open",
         {&BakaEngine::BakaOpen,
          {
           tr("[file]"),
           tr("opens the open file dialog or the file specified"),
           QString()
          }
         }
        },
        {"play_pause",
         {&BakaEngine::BakaPlayPause,
          {
           QString(),
           tr("toggle play/pause state"),
           QString()
          }
         }
        },
        {"fitwindow",
         {&BakaEngine::BakaFitWindow,
          {
           tr("[percent]"),
           tr("fit the window"),
           QString()
          }
         }
        },
        {"deinterlace",
         {&BakaEngine::BakaDeinterlace,
          {
           QString(),
           tr("toggle deinterlace"),
           QString()
          }
         }
        },
        {"interpolate",
         {&BakaEngine::BakaInterpolate,
          {
           QString(),
           tr("toggle motion interpolation"),
           QString()
          }
         }
        },
        {"mute",
         {&BakaEngine::BakaMute,
          {
           QString(),
           tr("mutes the audio"),
           QString()
          },
         }
        },
        {"volume",
         {&BakaEngine::BakaVolume,
          {
           tr("[level]"),
           tr("adjusts the volume"),
           QString()
          }
         }
        },
        {"speed",
         {&BakaEngine::BakaSpeed,
          {
           tr("[ratio]"),
           tr("adjusts the speed"),
           QString()
          }
         }
        },
        {"fullscreen",
         {&BakaEngine::BakaFullScreen,
          {
           QString(),
           tr("toggles fullscreen state"),
           QString()
          }
         }
        },
        {"hide_all_controls",
         {&BakaEngine::BakaHideAllControls,
          {
           QString(),
           tr("toggles hide all controls state"),
           QString()
          }
         }
        },
        {"boss",
         {&BakaEngine::BakaBoss,
          {
           QString(),
           tr("pause and hide the window"),
           QString()
          }
         }
        },
        {"clear",
         {&BakaEngine::BakaClear,
          {
           QString(),
           tr("clears the output textbox"),
           QString()
          }
         }
        },
        {"help",
         {&BakaEngine::BakaHelp,
          {
           tr("[command]"),
           tr("internal help menu"),
           QString()
          }
         }
        },
        {"about",
         {&BakaEngine::BakaAbout,
          {
           tr("[qt]"),
           tr("open about dialog"),
           QString()
          }
         }
        },
        {"msg_level",
         {&BakaEngine::BakaMsgLevel,
          {
           tr("[level]"),
           tr("set mpv msg-level"),
           QString()
          }
         }
        },
        {"quit",
         {&BakaEngine::BakaQuit,
          {
           QString(),
           tr("quit baka-mplayer"),
           QString()
          }
         }
        }
    };
    // Baka Command Functions
    void BakaMpv(QStringList&);
    void BakaSh(QStringList&);
    void BakaNew(QStringList&);
    void BakaOpenLocation(QStringList&);
    void BakaOpenClipboard(QStringList&);
    void BakaShowInFolder(QStringList&);
    void BakaAddSubtitles(QStringList&);
    void BakaAddAudio(QStringList&);
    void BakaScreenshot(QStringList&);
    void BakaMediaInfo(QStringList&);
    void BakaStop(QStringList&);
    void BakaPlaylist(QStringList&);
    void BakaJump(QStringList&);
    void BakaDim(QStringList&);
    void BakaOutput(QStringList&);
    void BakaPreferences(QStringList&);
    void BakaOnlineHelp(QStringList&);
    void BakaUpdate(QStringList&);
    void BakaOpen(QStringList&);
    void BakaPlayPause(QStringList&);
    void BakaFitWindow(QStringList&);
    void BakaAspect(QStringList&);
    void BakaDeinterlace(QStringList&);
    void BakaInterpolate(QStringList&);
    void BakaMute(QStringList&);
    void BakaVolume(QStringList&);
    void BakaSpeed(QStringList&);
    void BakaFullScreen(QStringList&);
    void BakaHideAllControls(QStringList&);
    void BakaBoss(QStringList&);
    void BakaClear(QStringList&);
    void BakaHelp(QStringList&);
    void BakaAbout(QStringList&);
    void BakaMsgLevel(QStringList&);
    void BakaQuit(QStringList&);
public:
    void Open();
    void OpenLocation();
    void Screenshot(bool subs);
    void MediaInfo(bool show);
    void PlayPause();
    void Jump();
    void FitWindow(int percent = 0, bool msg = true);
    void Dim(bool dim);
    void About(QString what = QString());
    void Quit();
};

#endif // BAKAENGINE_H
